Fix: config race conditions with RWMutex

EnameEtavir opened the pull request:
Added RWMutex to config.Instance which read or write locks
all instances where viper is used.

Refactored checksum manager to only use config and not
viper directly anymore.

All stash viper operations are now "behind" the config.Instance
and thus mutex "protected".

Partially resolves #1343. 
But for me those 2 concerns in #1343 should be separated. 
Concurrent writes are a bug, the other is an enhancement.

WithoutPants commented:
This does not fully resolve the race condition. I added a quick and dirty unit test:

```
func TestConcurrentConfigAccess(t *testing.T) {
	i := Instance{}

	const threads = 8
	const loops = 1000
	var wg sync.WaitGroup
	for t := 0; t < threads; t++ {
		wg.Add(1)
		go func() {
			for l := 0; l < loops; l++ {
				i.Set(CalculateMD5, !i.IsCalculateMD5())
			}
			wg.Done()
		}()
	}

	wg.Wait()
}
```

This crashed when I ran it. We'll need mutex protection on read and write operations.

EnameEtavir commented:
I should have done this in the first place. Thanks for also already providing your test code.
I've updated the PR.
Now all reads also use the RLock. Some small refactoring was done to not deadlock.
I have left in a library (commented out) used to more easily find deadlocks.

Also I've added the test you provided and extended it to nearly 100% of the methods in the config package.
(I did not change the value from read to write because for the concurrency testing the value itself should not relevant.)

While testing this specifically I also ran it with `-race`. I only found how you can exclude files from `-race` but not how to only run specific ones with `-race`.
Maybe it would be a good idea to add `-race` to the Makefile `test` ?
